### PR TITLE
fix(query) walking in MultiPartitionPlanner

### PR DESCRIPTION
**Pull request by amolnayak311**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** 

Tree walking was not happening to push the plan entirely to the planner causing  raw data to be pulled and aggregated.

**New behavior :**

The subtree is checked if it's in a single partition, if yes, entire plan is sent to the underlying planner (local or remote) else the operation is done in process.

### CODE REVIEW

File: coordinator/src/test/scala/filodb.coordinator/queryplanner/MultiPartitionPlannerSpec.scala [hunk 1]
```diff
@@ -1022,61 +1022,128 @@ class MultiPartitionPlannerSpec extends AnyFunSpec with Matchers {
...
+      |--E~BinaryJoinExec(binaryOp=ADD, on=List(), ignoring=List()) on InProcessPlanDispatcher(filodb.core.query.QueryConfig@298b64f7)
+      |---E~PromQlRemoteExec(PromQlQueryParams(sum(test3{job="app2"}),1000,100,10000,None,false), PlannerParams(filodb,None,None,None,30000,1000000,100000,100000,false,86400000,86400000,false,true,false,false), queryEndpoint=remote-url-1, requestTimeoutMs=10000) on InProcessPlanDispatcher(filodb.core.query.QueryConfig@298b64f7)
+      |---E~PromQlRemoteExec(PromQlQueryParams(sum(test4{job="app3"}),1000,100,10000,None,false), PlannerParams(filodb,None,None,None,30000,1000000,100000,100000,false,86400000,86400000,false,true,false,false), queryEndpoint=remote-url-2, requestTimeoutMs=10000) on InProcessPlanDispatcher(filodb.core.query.QueryConfig@298b64f7)""".stripMargin
+
+    validatePlan(execPlan, expectedPlan)
+
+  }
+
+
+  it("should materialize a multi level multi partition binary join correctly 2") {
```
Comment (vishramachandran) [inline coordinator/src/test/scala/filodb.coordinator/queryplanner/MultiPartitionPlannerSpec.scala:1054]: rename test to: should push multi-namespace portion of query when all of it is in one partition

File: coordinator/src/test/scala/filodb.coordinator/queryplanner/MultiPartitionPlannerSpec.scala [hunk 2]
```diff
@@ -1022,61 +1022,128 @@ class MultiPartitionPlannerSpec extends AnyFunSpec with Matchers {
...
+        |----T~AggregateMapReduce(aggrOp=Sum, aggrParams=List(), without=List(), by=List())
+        |-----T~PeriodicSamplesMapper(start=1000000, step=100000, end=10000000, window=None, functionId=None, rawSource=true, offsetMs=None)
+        |------E~MultiSchemaPartitionsExec(dataset=timeseries, shard=21, chunkMethod=TimeRangeChunkScan(700000,10000000), filters=List(ColumnFilter(job,Equals(app1)), ColumnFilter(__name__,Equals(test2))), colName=None, schema=None) on ActorPlanDispatcher(Actor[akka://default/system/testProbe-1#450220728],raw)
+        |-E~PromQlRemoteExec(PromQlQueryParams(ln((sum(test3{job="app2"}) + sum(test4{job="app2"}))),1000,100,10000,None,false), PlannerParams(filodb,None,None,None,30000,1000000,100000,100000,false,86400000,86400000,false,true,false,false), queryEndpoint=remote-url-1, requestTimeoutMs=10000) on InProcessPlanDispatcher(filodb.core.query.QueryConfig@298b64f7)""".stripMargin
+
+    validatePlan(execPlan, expectedPlan)
+  }
+
+  it("should materialize a multi level multi partition binary join correctly 3") {
```
Comment (vishramachandran) [inline coordinator/src/test/scala/filodb.coordinator/queryplanner/MultiPartitionPlannerSpec.scala:1113]: rename test to: should push entire query to remote partition when all of it is in one partition

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala [hunk 3]
```diff
@@ -61,43 +61,81 @@ class MultiPartitionPlanner(partitionLocationProvider: PartitionLocationProvider
       localPartitionPlanner.materialize(logicalPlan, qContext)
     }  else if (LogicalPlan.hasSubqueryWithWindowing(logicalPlan) || logicalPlan.isInstanceOf[TopLevelSubquery]) {
       materializeSubquery(logicalPlan, qContext)
-    } else {
-        walkLogicalPlanTree(logicalPlan, qContext).plans.head
+    } else logicalPlan match {
```
Comment (vishramachandran) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala:64]: Can you put the overall strategy as pseudocode in the javadoc of materialize method ? Could do for other planners as well - thank you for helping the next person who reads the planner code.

**Review by vishramachandran — APPROVED** [review]
Looks great. Thanks for the fixes.

Approved with some comments.

File: query/src/test/scala/filodb/query/PlanValidationSpec.scala [hunk 4]
```diff
@@ -7,16 +7,16 @@ import filodb.query.exec.ExecPlan
...
-   * Given an ExecPlan validate the provided plan string matches the required plab
+   * Given an ExecPlan validate the provided plan string matches the required plan
    *
    * @param plan LogicalPlan instance
    * @param expected expected plan as String
    */
   def validatePlan(plan: ExecPlan, expected: String): Unit = {
     val planString = plan.printTree()
-      .replaceAll("testProbe-.*\\]", "testActor]")
+      .replaceAll("testProbe-.*]", "testActor]")
```
[the remark this answers is not in the document]
Comment (amolnayak311) [reply query/src/test/scala/filodb/query/PlanValidationSpec.scala:17]: Yes, in fact IDEA recommended these changes and I changed them.

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala (hunk 3 above)
Comment (amolnayak311) [reply coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala:64]: Ah yes, I forgot that.. will do..

**Pull request merged**